pkg/control/sidecarcontrol/util.go

```diff
@@ -270,13 +274,101 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
 	}
 
 	for _, tEnv := range sidecarContainer.TransferEnv {
-		key := fmt.Sprintf("%v/%v", tEnv.SourceContainerName, tEnv.EnvName)
-		env, ok := envsInPod[key]
-		if !ok {
-			klog.Warningf("there is no env %v in container %v", tEnv.EnvName, tEnv.SourceContainerName)
-			continue
+		envs := sets.NewString()
+		if tEnv.EnvName != "" {
+			envs.Insert(tEnv.EnvName)
+		}
+		for _, e := range tEnv.EnvNames {
+			envs.Insert(e)
+		}
+
+		sourceContainerName := tEnv.SourceContainerName
+		if tEnv.SourceContainerNameFrom != nil && tEnv.SourceContainerNameFrom.FieldRef != nil {
+			containerName, err := ExtractContainerNameFromFieldPath(tEnv.SourceContainerNameFrom.FieldRef, pod)
+			if err != nil {
+				klog.Errorf("unmarshal pod(%s.%s) annotations[%s] failed: %s", pod.Namespace, pod.Name, err.Error())
+				continue
+			}
+			sourceContainerName = containerName
+		}
+		// if sourceContainerName is empty or not found in pod.spec.containers
+		for _, envName := range envs.List() {
+			key := fmt.Sprintf("%v/%v", sourceContainerName, envName)
+			env, ok := envsInPod[key]
+			if !ok {
+				klog.Warningf("there is no env %v in container %v", tEnv.EnvName, tEnv.SourceContainerName)
+				continue
+			}
+			injectedEnvs = append(injectedEnvs, env)
 		}
-		injectedEnvs = append(injectedEnvs, env)
 	}
 	return
 }
+
+func ExtractContainerNameFromFieldPath(fs *corev1.ObjectFieldSelector, pod *corev1.Pod) (string, error) {
+	fieldPath, _, err := ConvertDownwardAPIFieldLabel(fs.APIVersion, fs.FieldPath, "")
```

Comment:
Can this method refer directly to the K8S native method?

Reply:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/pods/helpers.go#L79-L93 the origin function supports more than we needed(metadata.annotations,metadata.labels,metadata.name)

Reply:
Lines 336-344 in the uitl.go file have been verified. Make sure we only support these fields, ok?

Reply:
yes,I think container name won't appear in pod status.